### PR TITLE
add RPC-style endpoint to submit and run a query

### PR DESCRIPTION
**Pull request by mathemancer**
This adds an endpoint `/api/db/v0/queries/run/` that takes a `POST` with a valid query definition, runs it, and returns an object with four keys:
- `query` is the query definition as given, plus any extra metadata added by the query serializer (e.g., the schema associated with the query's base table).
- `records` is an array containing the results of running the query (analogous to `/queries/<id>/records/`)
- `columns` is an array of column aliases.
- `column_metadata` is an object with column aliases mapping to metadata about the columns.

### Request format

The `POST` request body can contain the following keys:

- `initial_columns` -- same as persisted query `POST`
- `transformations` -- same as persisted query `POST` (optional)
- `base_table` -- same as persisted query `POST`, table ID
- `parameters` -- key-value map in the same form as query string parameters that could be submitted when `GET`ting the results of running a query (e.g., grouping, limit, offset, etc.) Optional.

Note that in contrast to the persisted query endpoint, no `name` key can be given, since that would be meaningless in this context, and potentially creates some confusion.

Example submission (this creates an overdue books report, with some additional filtering, ordering, and pagination to show how these work):

```json
{
  "initial_columns": [
    {"id": 19, "alias": "Checkout id"},
    {"id": 23, "alias": "Due"},
    {"id": 24, "alias": "Check In"},
    {"id": 8, "alias": "email", "jp_path": [[21, 5]], "display_name": "Patron Email"},
    {"id": 16, "alias": "Book Title", "jp_path": [[20, 1], [2, 13]]}
  ],
  "transformations": [
    {
      "spec": {"lesser": [{"column_name": ["Due"]}, {"literal": ["2022-08-10"]}]},
      "type": "filter"
    }, {
      "spec": {"empty": [{"column_name": ["Check In"]}]},
      "type": "filter"
    }, {
      "spec": {
        "grouping_expressions": [{"input_alias": "email"}],
        "aggregation_expressions": [
          {"function": "aggregate_to_array", "input_alias": "Book Title", "output_alias": "Title List"}
        ]
      },
      "type": "summarize",
      "display_names": {"Title List": "Titles"}
    }
  ],
  "base_table": 5,
  "parameters": {"order_by": [{"field": "email", "direction": "desc"}], "limit": 3, "offset": 1}
}
```

Change IDs as appropriate, of course. If one _were_ to submit actual query string parameters with their `POST` request, these will be merged with the `parameters` given in the `POST` body, with the latter taking precedence.

### Response format

The above query receives the following example response:
```json
{
    "query": {
        "schema": 1,
        "initial_columns": [
            {"id": 19, "alias": "Checkout id"},
            {"id": 23, "alias": "Due"},
            {"id": 24, "alias": "Check In"},
            {"id": 8, "alias": "email", "jp_path": [[21, 5]], "display_name": "Patron Email"},
            {"id": 16, "alias": "Book Title", "jp_path": [[20, 1], [2, 13]]}
        ],
        "transformations": [
            {
                "spec": {"lesser": [{"column_name": ["Due"]}, {"literal": ["2022-08-10"]}]},
                "type": "filter"
            }, {
                "spec": {"empty": [{"column_name": ["Check In"]}]},
                "type": "filter"
            }, {
                "spec": {
                    "grouping_expressions": [{"input_alias": "email"}],
                    "aggregation_expressions": [
                        {
                            "function": "aggregate_to_array",
                            "input_alias": "Book Title",
                            "output_alias": "Title List"
                        }
                    ]
                },
                "type": "summarize",
                "display_names": {"Title List": "Titles"}
            }
        ],
        "base_table": 5
    },
    "records": {
        "count": 7,
        "grouping": null,
        "preview_data": null,
        "results": [
            {"email": "tevans46@thompson.net", "Title List": ["Space Music Rest Crime"]},
            {"email": "kwright@odonnell.com", "Title List": ["On Letter Experience", "Day Beyond Property"]},
            {"email": "kathyb@le.org", "Title List": ["Pass Street Year"]}
        ]
    },
    "output_columns": ["email", "Title List"],
    "column_metadata": {
        "Checkout id": {
            "alias": "Checkout id", "display_name": null,
            "type": "integer", "type_options": null,
            "display_options": null
        },
        "Due": {
            "alias": "Due", "display_name": null,
            "type": "date", "type_options": null,
            "display_options": null
        },
        "Check In": {
            "alias": "Check In", "display_name": null,
            "type": "timestamp without time zone", "type_options": null,
            "display_options": null
        },
        "email": {
            "alias": "email", "display_name": "Patron Email",
            "type": "mathesar_types.email", "type_options": null,
            "display_options": null
        },
        "Book Title": {
            "alias": "Book Title", "display_name": null,
            "type": "text", "type_options": null,
            "display_options": null
        },
        "Title List": {
            "alias": "Title List", "display_name": "Titles",
            "type": "_array", "type_options": {"item_type": "text"},
            "display_options": null
        }
    },
    "parameters": {"order_by": [{"field": "email", "direction": "desc"}], "limit": 3, "offset": 1}
}
```

Beyond the description above, Note that the `column_metadata` objects each have the following keys:
- `alias` -- This is the identifier of the column, and is taken to be unique. It's also the key for the metadata object.
- `display_name` -- This can be submitted for initial columns, and added to in transformations. The last display name set takes precedence for a given alias in the response.
- `type` -- This is the type of a given column associated with an alias. In particular, in cases of more than one summarization step, this allows the UI to recover the database type of any column that has "existed" at any point in the query pipeline.
- `type_options` -- This is the normal type option value (e.g., string length, number precision, array type, etc.)
- `display_options` -- This is currently untested, but should hold any display options set for the initial column associated with a given alias. Right now, this is only working for the alias originally given to an initial column.

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by mathemancer** [comment]
@pavish The only review I need from you is to double-check that the request/response formats look good (though you're of course welcome to go more in-depth if you wish).

File: db/queries/base.py [hunk 1]
```diff
@@ -40,6 +40,32 @@ def get_count(self, **kwargs):
...
+            for col in DBQuery(
+                base_table_oid=self.base_table_oid,
+                initial_columns=self.initial_columns,
+                engine=self.engine,
+                transformations=self.transformations[:i],
+                name=f'{self.name}_{i}'
+            ).transformed_relation.columns
+        }
+
+        return initial_columns_map | transforms_columns_map | output_columns_map
```
[the remark this answers is not in the document]
Comment (mathemancer) [reply db/queries/base.py:67]: That's a good point.

**Comment by dmos62** [comment]
I wish there were more comments. Otherwise, looks good.

**Comment by pavish** [comment]
I see that the PR is merged in already.

@mathemancer 
I think the API spec looks good. We had already discussed the API spec on a previous call and it seems inline with what we had in mind.

File: db/queries/base.py [hunk 2]
```diff
@@ -40,6 +40,32 @@ def get_count(self, **kwargs):
...
+            col.name: MathesarColumn.from_column(col, engine=self.engine)
+            for i in range(len(self.transformations))
+            for col in DBQuery(
+                base_table_oid=self.base_table_oid,
+                initial_columns=self.initial_columns,
+                engine=self.engine,
+                transformations=self.transformations[:i],
+                name=f'{self.name}_{i}'
+            ).transformed_relation.columns
+        }
```
[the remark this answers is not in the document]
Comment (mathemancer) [reply db/queries/base.py:65]: That's true. I'll try to sneak that in via another PR.

**Comment by dmos62** [comment]
@pavish sorry for jumping the gun on the merge. I was under the false impression that my review was holding up the merge.

**Pull request merged**